### PR TITLE
[streamcz] backport from yt-dlp

### PR DESCRIPTION
**Pull request by arkamar**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This extractor originally started as https://github.com/ytdl-org/youtube-dl/pull/21774 PR that later evolved to https://github.com/yt-dlp/yt-dlp/pull/1616 which was merged to `yt-dlp`. This is attempt to backport it to `youtube-dl` with minimal changes as I suggested in https://github.com/ytdl-org/youtube-dl/pull/30462#issuecomment-1032477326. Tests are synchronized with PR https://github.com/yt-dlp/yt-dlp/pull/2747.

Fixes: #30160
Fixes: #29689
Fixes: #20900
Fixes: #22114

### CODE REVIEW

File: youtube_dl/extractor/streamcz.py [hunk 1]
```diff
@@ -1,105 +1,124 @@
...
+                yield {
+                    'format_id': '{}-{}'.format(format_id, ext),
+                    'ext': ext,
+                    'source_preference': pref,
+                    'url': urljoin(spl_url, stream['url']),
+                    'tbr': float_or_none(stream.get('bandwidth'), scale=1000),
+                    'duration': float_or_none(stream.get('duration'), scale=1000),
+                    'width': stream.get('resolution', 2 * [0])[0] or None,
+                    'height': stream.get('resolution', 2 * [0])[1] or int_or_none(format_id.replace('p', '')),
+                    **parse_codecs(stream.get('codec')),
```
[the remark this answers is not in the document]
Comment (pukkandan) [reply youtube_dl/extractor/streamcz.py:70]: `utils.merge_dicts`

[the remark this answers is not in the document]
Comment (arkamar) [reply youtube_dl/extractor/streamcz.py]: thanks!

**Comment by dirkf** [comment]
Great, thanks: looks good to go.

**Comment by arkamar** [comment]
thanks!

**Pull request merged**